File: linked_list/reverse_linked_list.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # 返回ListNode
    def ReverseList(self, pHead):
        if pHead == None:
            return None
        elif pHead.next == None:
            return pHead

        pre = None
        cur = pHead
        nex = None
        while cur:
            nex = cur.next
            cur.next = pre
            pre = cur
            cur = nex
        return pre

# Test
l1 = ListNode(1)
l1.next = ListNode(2)
l1.next.next = ListNode(3)
solution = Solution()
l = solution.ReverseList(l1)

print(l.val)
print(l.next.val)
print(l.next.next.val)

File: linked_list/reverse_linked_list.cpp
#include <iostream>

struct ListNode {
	int val;
	struct ListNode *next;
	ListNode(int x) :
			val(x), next(nullptr) {
	}
};

ListNode *reverseLinkedList(ListNode *pHead)
{
    if (pHead == nullptr)
    {
        return nullptr;
    }

    ListNode *pre, *cur, *nex;
    pre = nullptr;
    cur = pHead;
    nex = nullptr;

    while(cur)
    {
        nex = cur->next;
        cur->next = pre;
        pre = cur;
        cur = nex;
    }
    return pre;
}

int main()
{
    ListNode node1(3);
    ListNode node2(4);
    ListNode node3(5);
    node1.next = &node2;
    node2.next = &node3;
    ListNode *reversed = reverseLinkedList(&node1);
    std::cout << reversed->val << "; " <<  reversed->next->val << "; " << reversed->next->next->val << std::endl;
}
